Implement the Correlator special suite

JamesWrigley opened the pull request:
Finally got around to making a PR out of this :sweat_smile: I'm keeping it as a draft for now because there are a couple of (hopefully) minor issues still to fix:
- f581c873 needs to be refactored, you can ignore this commit for now.
- This uses the [`high_high_water_mark`](https://git.xfel.eu/karaboDevices/metropc/-/tree/high_high_water_mark) branch of `metropc`, which in turn is based on [`feat/bound_device_features`](https://git.xfel.eu/karaboDevices/metropc/-/tree/feat/bound_device_features). The `high_high_water_mark` branch just increases the HWM for the output socket of `metropc`, because I noticed that sometimes `metropc` was just dropping outputs and that fixed the issue. It could also be because the special suite is not reading fast enough, but that needs to be investigated. And perhaps we should wait to merge this until the `feat/bound_device_features` branch is merged too.

To reviewers, even though this is a draft I would suggest starting now because there's... a lot... of code. There's also some auxiliary commits that were needed for the Correlator, so I would suggest reviewing each commit individually because they're fairly atomic (it'll make more sense that way).

~~Note: this PR depends on #326, so you should ignore commits 1581c1b2 and 479e93a8.~~

JamesWrigley committed:
Implement the Correlator special suite

JamesWrigley committed:
Add tests for the Correlator

JamesWrigley committed:
Remove the 'category' abstraction from the special suites

JamesWrigley committed:
Implement SimpleSequence.resize()

JamesWrigley committed:
Change data_with_assembled() to save the train ID as "train_id"

This is to match what the DataTransformer does.

JamesWrigley committed:
Add documentation for the Correlator

JamesWrigley commented:
Hmm, it's failing because the CI can't install `metropc`. Currently it's using an SSH URL but the repo is private anyway so using HTTPS won't help. @philsmt, is there a timeline to make `metropc` public?

philsmt commented on extra_foam/utils.py:
```diff
@@ -277,3 +281,96 @@ def get_available_port(default_port):
+                               metadata for plotting.
+    """
+    xr_attrs = {
+        "xlabel": xlabel,
```
When receiving data as an `xarray.DataArray` (as you create here), there is already a documented [protocol](https://github.com/philsmt/metro-sci/blob/master/src/metro/devices/display/fast_plot.py#L1222) in `metropc` to use the coordinate label(s) as X axis label in 1D, whereas the y label is passed as an attribute (like here).

Would you mind using this instead of an `xlabel` attribute for compatibility?

JamesWrigley replied:
Yeah sure, that should be fine.

JamesWrigley committed:
Auto-level the first image set on an ImageViewF

Often the first thing a user does upon starting processing is auto-level all the
views in the window, and usually the intensity doesn't change dramatically from
the first image received so we might as well attempt to set it automatically for
them.

JamesWrigley committed:
Use slightly looser semantics for checking for ImageViewF's

This is useful for the Correlator so that its image view receives the auto-level
signals. However, this is not the cleanest implementation...

JamesWrigley committed:
Add support for labels to RectROI

JamesWrigley committed:
Add support for ROIs generated from View's

This is a very simple implementation of something that can hopefully be improved
and cleaned up in the future: support for context-defined annotation widgets.

Each ROI is defined as a parameter, they are added to a View with an argument to
the Image_WithRoi type, which is read by the Correlator and thus creates an
ROI. When the ROI is changed, an updated parameter is sent to metropc and the
Correlator modifies the context file with the new ROI parameters (position,
height, etc).

JamesWrigley committed:
Add the hostname/port/client type to the special suite interface

This makes it easier to change the settings from within special suites, without
having to dig into _SpecialAnalysisBase code.

JamesWrigley committed:
Add margin symbols for context errors

Also added a warning dialog that appears when the user tries to move a ROI when
there's a syntax error in the code (which libcst will fail to parse).

JamesWrigley committed:
WIP support for downsampling large vectors

JamesWrigley committed:
Add support for images with dynamic aspect ratios

JamesWrigley committed:
Show image axes by default

JamesWrigley committed:
WIP implementation of binning for motor scans

JamesWrigley committed:
Implement curve fitting for individual series

JamesWrigley committed:
Add an ElidedLabel class to display long context paths more neatly

JamesWrigley committed:
Allow having separate X values for each series

Among other things, this makes it possible to downsample all series being
plotted.

JamesWrigley committed:
Add support for specifying the beam width of error bars

JamesWrigley committed:
Make the ROI creation code more flexible

This is to allow LinearROIs on images.

JamesWrigley committed:
Allow setting ROIs on AspectUnlocked views

JamesWrigley committed:
Add a utility method for getting slice objects from LinearROI

JamesWrigley committed:
Add support for action views in the interface

JamesWrigley committed:
Implement a get_icon_path() helper

JamesWrigley committed:
Add live validation to the correlator editor

JamesWrigley committed:
Install PyQt libraries through conda

This is in preparation for using another library that's most up-to-date on
conda.

JamesWrigley committed:
Fix tests with newer packages

- Replace np.warnings with the warnings module
- Replace np.float with float
- Pass enum values instead of the enums themselves to redis functions
- Drop support for saving float arrays PNG images and expecting them to be
  restored with the same values. Seems that Pillow dropped support for it.

JamesWrigley committed:
Update correlator tests

Lots of minor fixes to the tests, and fixed one bug in the correlator that
caused an extra nan value to be added to the error arrays for each series.

philsmt commented:
We should probably aim to point this at `metropc`'s `main` branch 🤔

JamesWrigley commented:
I think you pressed the wrong button there :smile:

philsmt commented:
I did, apologies 😳

JamesWrigley committed:
fixup! Add live validation to the correlator editor

takluyver committed:
Merge branch 'dev' into correlator

takluyver committed:
wget doesn't like downloading to path containing ..

takluyver committed:
Update metropc to master branch
